src/examples/zkapps/voting/test.ts

```diff
@@ -28,46 +41,769 @@ export async function testSet(
     votersStore: Voters;
   }
 ) {
-  let Local = Mina.LocalBlockchain();
-  Mina.setActiveInstance(Local);
+  let { votersStore, candidatesStore, votesStore } = storage;
+  let { votingKey } = params;
 
-  let feePayer = Local.testAccounts[0].privateKey;
+  /*
+    test case description:
+      voting contract is trying to call methods of an invalid contract 
+    
+    preconditions:
+      - real voting contract deployed
+      - voter and candidate membership contracts are faulty (empty/dummy contracts)
+      - trying to register a valid voter member
 
-  let tx;
+    tested cases:
+      - deploying set of invalid contracts
+        - trying to invoke a non-existent method -> failure
 
-  let voterContract = contracts.voterContract;
-  let candidateContract = contracts.candidateContract;
-  let voting = contracts.voting;
+    expected results:
+      - throws an error
 
-  console.log('deploying set of 3 contracts');
-  tx = await Mina.transaction(feePayer, () => {
-    Party.fundNewAccount(feePayer, {
-      initialBalance: Mina.accountCreationFee().add(Mina.accountCreationFee()),
+  */
+
+  console.log('deploying testing phase 1 contracts');
+
+  let invalidSet = await deployInvalidContracts(
+    contracts,
+    params,
+    votersStore.getRoot(),
+    candidatesStore.getRoot(),
+    votesStore.getRoot()
+  );
+
+  console.log('trying to invoke invalid contract method...');
+
+  try {
+    let tx = await Mina.transaction(invalidSet.feePayer, () => {
+      let m = Member.from(
+        PrivateKey.random().toPublicKey(),
+        Field.zero,
+        UInt64.from(15)
+      );
+      invalidSet.Local.addAccount(m.publicKey, m.balance.toString());
+
+      invalidSet.voting.voterRegistration(m);
+
+      invalidSet.voting.sign(votingKey);
     });
 
-    voting.deploy({ zkappKey: params.votingKey });
-    candidateContract.deploy({ zkappKey: params.candidateKey });
-    voterContract.deploy({ zkappKey: params.voterKey });
+    tx.send();
+  } catch (err: any) {
+    if (!err.toString().includes('precondition_unsatisfied')) {
+      throw Error('Transaction should have failed but went through!');
+    }
```

Comment:
We should also handle the cases when the transaction fails for any other reason than the expected failure case of 'precondition_unsatisfied'.

    if (!err.toString().includes('precondition_unsatisfied')) {
      throw Error('Transaction should have failed but went through!');
    } else {
      // Throw an error on all unexpected failure cases
      throw Error(error)
    }

Reply:
good point!